Engine/src/Platform/OpenGL41/OpenGL41RenderTarget.cpp

```diff
@@ -0,0 +1,99 @@
+#include "Platform/OpenGL41/OpenGL41RenderTarget.h"
+#include "Platform/OpenGL41/OpenGL41Check.h"
+
+#include "Renderer/RHI/Texture.h"
+#include "Resources/TextureManager.h"
+
+#include <glad/gl.h>
+
+namespace elv {
+OpenGL41RenderTarget::~OpenGL41RenderTarget()
+{
+    glCheck(glDeleteFramebuffers(1, &m_id));
+}
+
+void OpenGL41RenderTarget::Init(const std::uint32_t width, const std::uint32_t height, bool isMultisampled)
+{
+    m_isMultisampled = isMultisampled;
+    m_size = { width, height };
+
+    glCheck(glGenFramebuffers(1, &m_id));
+    glCheck(glBindFramebuffer(GL_FRAMEBUFFER, m_id));
+    InitAttachments();
+    glBindFramebuffer(GL_FRAMEBUFFER, 0);
+}
+
+void OpenGL41RenderTarget::Bind() const
+{
+    glCheck(glBindFramebuffer(GL_FRAMEBUFFER, m_id));
+}
+
+void OpenGL41RenderTarget::Unbind() const
+{
+    glCheck(glBindFramebuffer(GL_FRAMEBUFFER, 0));
+}
+
+void OpenGL41RenderTarget::BindDraw() const
+{
+    glCheck(glBindFramebuffer(GL_DRAW_FRAMEBUFFER, m_id));
+}
+
+void OpenGL41RenderTarget::BindRead() const
+{
+    glCheck(glBindFramebuffer(GL_READ_FRAMEBUFFER, m_id));
+}
+
+void OpenGL41RenderTarget::Resize(const std::uint32_t width, const std::uint32_t height)
+{
+    m_size = { width, height };
+
+    Bind();
+    InitAttachments();
+    glBindFramebuffer(GL_FRAMEBUFFER, 0);
+}
+
+void OpenGL41RenderTarget::Blit(const UniquePtr<RenderTarget>& src)
+{
+    src->BindRead();
+    BindDraw();
+    glCheck(glBlitFramebuffer(0, 0, m_size.Width, m_size.Height, 0, 0, src->GetSize().Width, src->GetSize().Height, GL_COLOR_BUFFER_BIT, GL_NEAREST));
+}
+
+void OpenGL41RenderTarget::InitAttachments()
+{
+    {
+        Texture::Info info;
+
+        m_colorTextureAttachment = textures::Load(m_size.Width, m_size.Height, info);
+        glCheck(
+            glFramebufferTexture2D(
+                GL_FRAMEBUFFER,
+                GL_COLOR_ATTACHMENT0,
+                GL_TEXTURE_2D,
```

Comment:
Try to write `m_isMultisampled ? GL_TEXTURE_2D_MULTISAMPLE : GL_TEXTURE_2D,` here and set `info.Multisample`. It works for me for OpenGL41 API. If it doesn't work on Mac, then I think we can skip MSAA for opengl41.

Reply:
AFAIK, I have tried exactly that and didn't succeed. Although the pre-debugcallback error messages are not very descriptive (and not always documented), so maybe I missed something - I will try again, and if not, no MSAA would be available for opengl41 for now.